server/src/main/java/org/elasticsearch/search/SearchService.java

```diff
@@ -1596,6 +1608,37 @@ private CanMatchShardResponse canMatch(ShardSearchRequest request, boolean check
         }
     }
 
+    /**
+     * This method tries to rewrite a query without using a searcher (before we get a valid one). It takes advantage of the fact that
+     * we can skip some shards in the query phase because we have enough information in the index mapping to decide the 'can match'
+     * outcome. One such example are queries like {@link org.elasticsearch.index.query.MatchPhraseQueryBuilder} queries which are rewritten
+     * to {@link org.elasticsearch.index.query.MatchNoneQueryBuilder} after being rewritten to
+     * {@link org.elasticsearch.index.query.TermQueryBuilder}. Other queries have a similar pattern, especially when using
+     * {@link org.elasticsearch.index.mapper.ConstantFieldType} or unmapped fields, they all rewrite to
+     * {@link org.elasticsearch.index.query.MatchNoneQueryBuilder} (if conditions are met). This allows us to avoid extra work other than
+     * making the shard search active and waiting for refreshes. As a result, we only wait for refreshes to happen on shards that have
+     * actual data.
+     */
+    private static CanMatchShardResponse canMatchAfterRewriteWithoutSearcher(
+        final ShardSearchRequest request,
+        final IndexService indexService
+    ) throws IOException {
+        final SearchExecutionContext searchExecutionContext = new SearchExecutionContext(
+            indexService.newSearchExecutionContext(
+                request.shardId().getId(),
+                request.shardRequestIndex(),
+                null,
+                request::nowInMillis,
+                request.getClusterAlias(),
+                request.getRuntimeMappings()
+            )
+        );
+        if (queryStillMatchesAfterRewrite(request, searchExecutionContext) == false) {
+            return new CanMatchShardResponse(false, null);
+        }
+        return null;
```

Comment:
Perhaps this should return a boolean. If true we will go ahead, otherwise we will return a can match shard response with the flag set to false?